packages/daimo-api/src/contract/ClogMatcher.ts

```diff
@@ -0,0 +1,309 @@
+      // Start event is an outbound cross-chain transfer, FastFinish is inbound
+      // const outbound = row.log_name === "Start";
+      // const tokenAddr = outbound ?  : assertNotNull(row.from_token);
+      // const crossChainId = outbound ?  : row.from_chain_id;
+
+      const fastCctpLog = {
+        blockHash: bytesToHex(row.block_hash, { size: 32 }),
+        blockNumber: BigInt(row.block_num),
+        transactionHash: bytesToHex(row.tx_hash, { size: 32 }),
+        transactionIndex: Number(row.tx_idx),
+        logIndex: Number(row.log_idx),
+        address: getAddress(bytesToHex(row.to_token, { size: 20 })),
+        from: getAddress(
+          bytesToHex(assertNotNull(row.from_addr), { size: 20 })
+        ),
+        to: getAddress(bytesToHex(row.to_addr, { size: 20 })),
+        value: BigInt(row.to_amount),
+      };
+
+      const crossChainToken = this.tokenReg.getToken(
+        fastCctpLog.address,
+        Number(row.to_chain_id), // cross-chain (to chain)
+        true
+      );
+      if (crossChainToken == null) continue;
+
+      const crossChainTransfer: ForeignTokenTransfer = {
+        ...fastCctpLog,
+        foreignToken: crossChainToken,
+      };
+      this.txHashToTransfers.set(crossChainTransfer.transactionHash, [
+        ...(this.txHashToTransfers.get(crossChainTransfer.transactionHash) ||
+          []),
+        crossChainTransfer,
+      ]);
+    }
+
+    const elapsedMs = (Date.now() - startTime) | 0;
+    console.log(
+      `[SWAP] loaded ${transferResult.length} bundled transfers and ${startResult.length} Start FastCCTP events in ${elapsedMs}ms`
+    );
+  }
+
+  /**
+   * For inbound cross-chain transfers, we need to fetch the "from token"
+   * information from the cross-chain Start event.
+   */
+  async loadCrossChainClogs(
+    kdb: Kysely<IndexDB>,
+    from: number,
+    to: number,
+    chainId: number,
+    blockNums: Set<bigint>,
+    txHashes: Set<Hex>
+  ) {
+    const startTime = Date.now();
+
+    const blockList = [...blockNums].sort().map((bn) => bn.toString());
+    const txList = [...txHashes].map((tx) => Buffer.from(hexToBytes(tx)));
+
+    const fastFinishResult = await retryBackoff(
+      `fastCctp-logs-query-${from}-${to}`,
+      () =>
+        kdb
+          .selectFrom("index.daimo_fast_cctp")
+          .selectAll()
+          .where("chain_id", "=", "" + chainId)
+          .where("block_num", "in", blockList)
+          .where("tx_hash", "in", txList)
+          .where("log_name", "=", "FastFinish")
+          .execute()
+    );
+
+    // For each FastFinish event, load the corresponding Start event
+    for (const row of fastFinishResult) {
+      const startLog = await retryBackoff(
+        `StartFastCctp-logs-query-${from}-${to}`,
+        () =>
+          kdb
+            .selectFrom("index.daimo_fast_cctp")
+            .selectAll()
+            .where("chain_id", "=", row.from_chain_id)
+            .where("handoff_addr", "=", row.handoff_addr)
```

Comment:
two thoughts:
- we can do this in one query (vs 1+n queries) with join on chain ID & handoff addr
- add an index on handoff_addr